docs/sequence.puml

```diff
@@ -4,52 +4,53 @@ actor Bob
 Actor Alice
 database Monero
 
-Bob --> Alice: Establish connection
+group Negotiation
 
-group Request Response Channel
-    Bob -> Alice: Request Amounts
-    note left: Btc Amount
+    Bob --> Alice: Establish connection
 
-    Alice --> Bob: Send Amounts
-    note right: Btc & Xmr Amounts
+    group Request Response Channel
+        Bob -> Alice: Swap request
+        note left: Btc Amount\naddress hints
+        note right: Alice stores peer id & address hints\nto contact Bob
+
+        Alice --> Bob: Swap response
+        note right: ACK or NACK\n Approx. XMR amount
```

Comment:
At this stage I don't understand why it is still an approximate amount. I understand that the amount on a SP website has to be approximate, but right before the swap I suppose it would be the actual amount this is agreed upon. 
Is there no agreement needed to the exact amounts of the swap prior to the execution? The amounts are not part of calculating tx proofs?

Reply:
I will have to verify by trying to code it but as far as I can see for Monero, the amount is not needed until Monero is actually locked. Maybe I am incorrect, I'll double check tomorrow.

We want Alice to be able to provide the amount at the last minute as otherwise, Bob can wait before locking Btc (at now cost) to see whether the trade will go in his favour.

I think this solution is more elegant than putting in place arbitrary time outs to prevent that.

Reply:
Is the approximation necessary because of the funding step for the Bitcoin on the user side? 
Are we planning to just trigger the actual swap once the Bitcoin funds arrived in the swap-tool wallet?

Reply:
Imho Alice should tell Bob the **exact** XMR amount she is willing to commit. 
If Bob is not happy with this amount (rate), he can still bail out before publishing a single transaction. 
Obviously, the amounts are not enforced in anyway and Alice could use a different amount at the time of funding. 
But this would be really bad, because now both parties have issued a transaction on-chain, and the fees are wasted.

Reply:
> Imho Alice should tell Bob the **exact** XMR amount she is willing to commit.
> If Bob is not happy with this amount (rate), he can still bail out before publishing a single transaction.
> Obviously, the amounts are not enforced in anyway and Alice could use a different amount at the time of funding.
> But this would be really bad, because now both parties have issued a transaction on-chain, and the fees are wasted.
> 
> // Edit //
> 
> Different said: Bob should know how much XMR he will receive **before** issuing the first transaction (`tx_fund`).

As a single market maker, I want to be sure that I provide the right rate to not loose money.
If you look at the current ecosystem, whether it's crypto or forex, single market maker always provide an approximate rate. Then the user sends money and the rate gets locked once the money is received by the service provider.
Some single market makers also provide a "lock rate" service where the rate is guaranteed but is generally higher than the estimated flexible rate.

We could do fixed rates now but I worried that when we want to make the single market maker software MVP ready and we start to look at the functionalities that potential operators would want, we will have to revert to a flexible approximative rate.

Reply:
At this stage, I am happy to roll with a fixed rate and defer the discussion to when we try to make the Service Provider MVP.

Reply:
> As a single market maker, I want to be sure that I provide the right rate to not loose money.
If you look at the current ecosystem, whether it's crypto or forex, single market maker always provide an approximate rate. Then the user sends money and the rate gets locked once the money is received by the service provider.
> Some single market makers also provide a "lock rate" service where the rate is guaranteed but is generally higher than the estimated flexible rate.

For centralized services (SP) this works well, because the trader cannot change its mind half way through. The trader sends the money and hopes to receive at least something in return. If this is not good enough, the SP reputation suffers.

Now I think the story is different: 
The trader receives a promise from the maker to execute the trade at a certain rate. If the maker does not go through with the promised rate, he is in breach if this promise. 

Since we can't enforce this promise, we can argue whether this promise should never have been made in the first place and use an approximate rate instead. 
Personally, I prefer to know exactly how much money I will receive. 

In my head this preferably because of this use case: 
Bob receives a rate from Alice and locks up funds on Bitcoin. 
Alice locks up a different amount on XMR and now wants Bob to co-sign `tx_redeem` for Bitcoin.
If the XMR amount is not _good enough_ he will simply decline and both parties lost transaction fees. Hence, Alice best chance to go through with the trade is to tell Bob in advance how much she is willing to commit.

Reply:
> At this stage, I am happy to roll with a fixed rate and defer the discussion to when we try to make the Service Provider MVP.

Didn't see your reply. Eitherway, I don't think it's a big issue to change later one.